java-client/src/main/java/co/elastic/clients/elasticsearch/_helpers/bulk/BulkIngester.java

```diff
@@ -306,42 +348,80 @@ public void flush() {
                     request = null;
                 }
 
-                return new RequestExecution<>(id, request, requestContexts, result);
+                return new RequestExecution<>(id, request, contexts, result);
             });
 
         if (exec != null) {
             // A request was actually sent
             exec.futureResponse.handle((resp, thr) -> {
                 if (resp != null) {
-                    // Success
-                    if (listener != null) {
-                        listenerInProgressCount.incrementAndGet();
-                        scheduler.submit(() -> {
-                            try {
-                                listener.afterBulk(exec.id, exec.request, exec.contexts, resp);
-                            }
-                            finally {
-                                if(listenerInProgressCount.decrementAndGet() == 0){
-                                    closeCondition.signalIfReady();
-                                }
+
+                    // Success? Checking if total or partial
+                    List<BulkResponseItem> failedRequestsCanRetry = resp.items().stream()
+                        .filter(i -> i.error() != null && i.status() == 429)
+                        .collect(Collectors.toList());
+
+                    if (failedRequestsCanRetry.isEmpty() || backoffPolicy.equals(BackoffPolicy.noBackoff())) {
+                        // Total success! ...or there's no retry policy implemented. Either way, can call
+                        listenerAfterBulkSuccess(resp, exec);
+                    } else {
+                        // Partial success, retrying failed requests if policy allows it
+                        // Keeping list of retryable requests/responses, to exclude them for calling
+                        // listener later
+                        List<BulkOperationRepeatable<Context>> retryableReq = new ArrayList<>();
+                        List<BulkOperationRepeatable<Context>> refires = new ArrayList<>();
+                        List<BulkResponseItem> retryableResp = new ArrayList<>();
+
+                        for (BulkResponseItem bulkItemResponse : failedRequestsCanRetry) {
+                            int index = resp.items().indexOf(bulkItemResponse);
+                            selectingRetries(index, bulkItemResponse, sentRequests, retryableResp,
+                                retryableReq, refires);
+                        }
+                        // Scheduling flushes for just sent out retryable requests
+                        if (!refires.isEmpty()) {
+                            scheduleRetries(refires);
+                        }
+                        // Retrieving list of remaining successful or not retryable requests
+                        sentRequests.removeAll(retryableReq);
```

Comment:
Avoid the costly `removeAll`. This can be refactored to use the less costly `Collectors.groupBy`